tests/test_helper_functions.py

```diff
@@ -0,0 +1,148 @@
+        self.assertFalse(helper_functions.has_mixed_signs(torch.tensor([-1])))
+
+        # Zeros are unsigned
+        self.assertFalse(helper_functions.has_mixed_signs(torch.tensor([0, 0])))
+        self.assertFalse(helper_functions.has_mixed_signs(torch.tensor([0, 1.2])))
+        self.assertFalse(helper_functions.has_mixed_signs(torch.tensor([0, -1.2])))
+
+        # All entries have same sign
+        self.assertFalse(helper_functions.has_mixed_signs(torch.tensor([0.4, 1.2])))
+        self.assertFalse(helper_functions.has_mixed_signs(-torch.tensor([0.4, 1.2])))
+
+        # Finally!
+        self.assertTrue(helper_functions.has_mixed_signs(torch.tensor([-0.1, 1.2])))
+
+    def test_bounded_in_plus_minus_one(self):
+        # Violation on one end
+        self.assertFalse(helper_functions.bounded_in_plus_minus_one(torch.tensor([1.2])))
+        self.assertFalse(helper_functions.bounded_in_plus_minus_one(torch.tensor([-1.2])))
+        self.assertFalse(helper_functions.bounded_in_plus_minus_one(torch.tensor([1.2, 0])))
+        self.assertFalse(helper_functions.bounded_in_plus_minus_one(torch.tensor([-1.2, 0])))
+
+        # Boundary
+        self.assertTrue(helper_functions.bounded_in_plus_minus_one(torch.tensor([1])))
+        self.assertTrue(helper_functions.bounded_in_plus_minus_one(torch.tensor([-1])))
+        self.assertTrue(helper_functions.bounded_in_plus_minus_one(torch.tensor([1, -1])))
+        self.assertTrue(helper_functions.bounded_in_plus_minus_one(torch.tensor([1, 0])))
+        self.assertTrue(helper_functions.bounded_in_plus_minus_one(torch.tensor([0, 1])))
+
+        # Correct
+        self.assertTrue(helper_functions.bounded_in_plus_minus_one(torch.tensor([0.5, 0.9, -0.2])))
+
+    @parameterized.expand([[dict(a=1, x=4)], [dict(a="hello")]])
+    def test_has_correct_config_common(self, kwargs):
+        class MyModel(torch.nn.Module):
+            @store_config
+            def __init__(self, a, b=2, *, x=4, y=5):
+                super().__init__()
+
+            def forward(self, x):
+                return torch.ones(5)
+        model = MyModel(**kwargs)
+        self.assertTrue(helper_functions.has_correct_config(model))
+
+    def test_has_correct_config(self):
+        self.assertFalse(helper_functions.has_correct_config(torch.nn.Linear(5, 3)))
+
+    def test_shapes_match(self):
+        shape = (123, 456)
+        x = torch.randn(shape)
+        self.assertTrue(helper_functions.shapes_match(x, shape))
+        self.assertFalse(helper_functions.shapes_match(x, (1, 2, 3)))
+
+    def test_model_probably_good(self):
+        with self.subTest("Model should be good"):
+            class MyModel(torch.nn.Module):
+                @store_config
+                def __init__(self, a, b=2, *, x=4, y=5):
+                    super().__init__()
+
+                def forward(self, x):
+                    return 2*x
+            self.assertTrue(helper_functions.model_probably_good(MyModel("hello"), (500, ), (500,)))
```

Comment:
To a point made earlier, this test could be simplified a bit:
```suggestion
        class MyModel(torch.nn.Module):
            @store_config
            def __init__(self, a, b=2, *, x=4, y=5):
                super().__init__()

        with self.subTest("Model should be good"):
            MyModel.forward = lambda self, x: 2 * x
            self.assertTrue(helper_functions.model_probably_good(MyModel("hello"), (500, ), (500,)))
```
but I'll leave this just as an option. Up to you whether you prefer this or the former.